### PR TITLE
Support OpenCV 4

### LINKED ISSUE #1672: YARP does not compile with OpenCV 4
I pulled `opencv` and `yarpdataplayer` and `yarpdatadumper` doesn't compile:
```
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:418:30: error: use of undeclared identifier 'IplImage'
                int frameW=((IplImage*)itemEnd.obj->getPtr())->width;
                             ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:418:39: error: expected expression
                int frameW=((IplImage*)itemEnd.obj->getPtr())->width;
                                      ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:419:30: error: use of undeclared identifier 'IplImage'
                int frameH=((IplImage*)itemEnd.obj->getPtr())->height;
                             ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:419:39: error: expected expression
                int frameH=((IplImage*)itemEnd.obj->getPtr())->height;
                                      ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:428:52: error: use of undeclared identifier 'CV_FOURCC'
                videoWriter.open(videoFile.c_str(),CV_FOURCC('H','F','Y','U'),
                                                   ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:429:38: error: use of undeclared identifier 'cvSize'
                                 fps,cvSize(frameW,frameH),true);
                                     ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:457:49: error: use of undeclared identifier 'IplImage'
                    cv::Mat img=cv::cvarrToMat((IplImage*)item.obj->getPtr());
                                                ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:457:58: error: expected expression
                    cv::Mat img=cv::cvarrToMat((IplImage*)item.obj->getPtr());
                                                         ^
/home/ngenesio/yarp/src/yarpdatadumper/main.cpp:457:37: error: no member named 'cvarrToMat' in namespace 'cv'
                    cv::Mat img=cv::cvarrToMat((IplImage*)item.obj->getPtr());

```

my `opencv_version` is `4.0.0-pre`.
I'm afraid that in opencv 4 they will remove `IplImage` and `cvarrToMat`.
We should make `yarp::sig::Image` compatible with `cv::Mat` without taking account of `IplImage`.

### PR DESCRIPTION
**Pull request by Nicogene**
It fixes #1672.

It compiles also with OpenCV 3

Please review code.

### CODE REVIEW

File: src/libYARP_cv/src/yarp/cv/Cv.h [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #include <type_traits>
 #include <opencv2/core/core.hpp>
 #include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/imgproc/types_c.h>
```
Comment (traversaro) [inline src/libYARP_cv/src/yarp/cv/Cv.h]: This header seems to be unused in this header, probably we can just add it where it is needed?

Comment (Nicogene) [reply src/libYARP_cv/src/yarp/cv/Cv.h]: I have to check, probably you are right, it is already included by `<opencv2/imgproc/imgproc.hpp`

Comment (Nicogene) [reply src/libYARP_cv/src/yarp/cv/Cv.h]: It is used by `Cv-inl.h` that it is included in the end of the file. I can move it there

**Comment by traversaro** [comment]
As the change is ABI-compatible and not impacting the public libraries at all (except the additional included mentioned in https://github.com/robotology/yarp/pull/2162#discussion_r360425391), could it make sense to backport it to `yarp-3.3`, so that for the Ubuntu 20.04 we will have a stable YARP version released, and we do not need to rush a release of YARP 3.4 in March/April?

**Comment by Nicogene** [comment]
> As the change is ABI-compatible and not impacting the public libraries at all (except the additional included mentioned in #2162 (comment)), could it make sense to backport it to yarp-3.3, so that for the Ubuntu 20.04 we will have a stable YARP version released, and we do not need to rush a release of YARP 3.4 in March/April?

Ok for me I can change the target branch

**Comment by traversaro** [comment]
> Ok for me I can change the target branch



> Ok for me I can change the target branch

What do you think @drdanz

**Comment by drdanz** [comment]
Debian stretch (which is compatible with YARP 3.3) does not have OpenCV3, therefore, unless this is compatible with OpenCV2, we should have it in master, or use `#ifdef` everywhere to support all 3 versions. Just 3 and 4 is OK on master, though

File: src/libYARP_cv/src/yarp/cv/Cv.h (hunk 1 above)
Comment (traversaro) [reply src/libYARP_cv/src/yarp/cv/Cv.h]: Ah ok, I had missed that.

**Comment by traversaro** [comment]
Sorry, I missed that this PR was breaking compatibility with OpenCV2 . If with some #ifdef magic it could be possible to support both OpenCV 2 and 4, it would be great at least for the releases YARP 3.3/3.4, then probably we can drop OpenCV 2 support for cleaning the code.

**Comment by Nicogene** [comment]
> Debian stretch (which is compatible with YARP 3.3) does not have OpenCV3, therefore, unless this is compatible with OpenCV2

On appveyor there is OpenCV `2.4.13` and it fails, then we should add #ifdefs or drop OpenCV 2.

E.g. https://ci.appveyor.com/project/robotology/yarp/builds/29681433/job/iynkchf54oxhp3b4

**Comment by Nicogene** [comment]
But the latest `icub-main` installer which version of opencv installs?
Because it is used to install opencv on `appveyor`. Because if we ship opencv 2, in case we want drop it we should change the installer too.

**Comment by traversaro** [comment]
For reference, the AppVeyor error is: 
> [00:29:37] C:\projects\yarp\src\yarpdataplayer\src\worker.cpp(30): fatal error C1083: Cannot open include file: 'opencv2/imgcodecs/imgcodecs.hpp': No such file or directory [C:\projects\yarp\build\src\yarpdataplayer\yarpdataplayer.vcxproj]

**Comment by Nicogene** [comment]
Now the PR makes `yarp` compatible with `opencv` 2, 3 and 4.

The only doubt I have, is that if https://github.com/robotology/yarp/pull/2162/commits/c5a88853b4f5dfc878b781f61c4a5a0a32d98831 is fine (cc @drdanz @traversaro).

>As the change is ABI-compatible and not impacting the public libraries at all (except the additional included mentioned in #2162 (comment)), could it make sense to backport it to yarp-3.3, so that for the Ubuntu 20.04 we will have a stable YARP version released, and we do not need to rush a release of YARP 3.4 in March/April?

The base branch has been changed to `yarp3.3`.
Let's see if appveyor and travis both pass.  

cc @pattacini @vtikha @maggia80

File: src/libYARP_cv/src/yarp/cv/Cv.h [hunk 2]
```diff
@@ -14,6 +14,9 @@
 #include <type_traits>
 #include <opencv2/core/core.hpp>
 #include <opencv2/imgproc/imgproc.hpp>
+#if not defined(YARP_HAS_OPENCV2)
```
Comment (traversaro) [inline src/libYARP_cv/src/yarp/cv/Cv.h]: Can't we just change this to: 
```suggestion
#if CV_MAJOR_VERSION >= 3
```
so we don't need to make sure that `YARP_HAS_OPENCV2` is correctly propagated to compilation units in downstream projects that use `libYARP_cv` ?

**Comment by traversaro** [comment]
> The only doubt I have, is that if [5f376a5](https://github.com/robotology/yarp/commit/5f376a550387ab5bdebe538dc97781b2f5d6bc03) is fine (cc @drdanz @traversaro).

Unfortunately as it is it will not work, as `YARP_HAS_OPENCV2` is not defined for compilation units of downstream projects that include `yarp/cv/Cv.h` from `libYARP_cv` . Can https://github.com/robotology/yarp/pull/2162/files#r362766966 be a valid alternative?

File: src/libYARP_cv/src/yarp/cv/Cv.h (hunk 2 above)
Comment (Nicogene) [reply src/libYARP_cv/src/yarp/cv/Cv.h]: For me it is ok :+1:

**Comment by Nicogene** [comment]
In the current state, on travis-macOs builds there is opencv 4, on travis-linux opencv 3, on appveyor opencv 2.

Then if everything passes the work can be considered done.

File: src/yarpdatadumper/main.cpp [hunk 3]
```diff
@@ -29,7 +29,11 @@
 #ifdef ADD_VIDEO
     #include <opencv2/opencv.hpp>
     #include <yarp/cv/Cv.h>
-#endif
+#if CV_MAJOR_VERSION >= 3
```
Comment (pattacini) [inline src/yarpdatadumper/main.cpp]: @Nicogene I would better indent this inner if-statement wrt the outer if-statement.

**Comment by drdanz** [comment]
Merged, thanks.

**Pull request merged**